EE6347 Assignment 5

Kishore Rajendran [EE20B064]

Part 1 : MNIST Classification using Pytorch

In [ ]:
# Importing Required Libraries
import torch
import torch.nn as nn
import torch.optim as optim
from torchvision import datasets, transforms
!pip install aihwkit

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 11.3/11.3 MB 82.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 619.9/619.9 MB 2.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 294.4/294.4 kB 28.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 21.0/21.0 MB 63.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 849.3/849.3 kB 64.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 11.8/11.8 MB 97.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 557.1/557.1 MB 2.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 317.1/317.1 MB 4.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 168.4/168.4 MB 6.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 54.6/54.6 MB 10.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 102.6/102.6 MB 9.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 173.2/173.2 MB 6.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━

In [ ]:
# Step 1: Getting dataset ready
transform = transforms.Compose([transforms.ToTensor(), transforms.Normalize((0.5,), (0.5,))])

# Importing the MNIST dataset
train_dataset = datasets.MNIST('data/', train=True, transform=transform, download=True)
test_dataset = datasets.MNIST('data/', train=False, transform=transform, download=True)

# Creating Data Loaders
batch_size = 100
train_loader = torch.utils.data.DataLoader(train_dataset, batch_size=batch_size, shuffle=True)
test_loader = torch.utils.data.DataLoader(test_dataset, batch_size=batch_size, shuffle=False)

100%|██████████| 9912422/9912422 [00:00<00:00, 95215840.81it/s]


Extracting data/MNIST/raw/train-images-idx3-ubyte.gz to data/MNIST/raw



100%|██████████| 28881/28881 [00:00<00:00, 21304202.22it/s]


Extracting data/MNIST/raw/train-labels-idx1-ubyte.gz to data/MNIST/raw



100%|██████████| 1648877/1648877 [00:00<00:00, 24014429.01it/s]


Extracting data/MNIST/raw/t10k-images-idx3-ubyte.gz to data/MNIST/raw



100%|██████████| 4542/4542 [00:00<00:00, 18806050.12it/s]


Extracting data/MNIST/raw/t10k-labels-idx1-ubyte.gz to data/MNIST/raw



Defining the NN architecture using fully connected layers with input sizes matching the MNIST dataset.

We use the softmax activation function in the last layer, since it's a multi-class classfifcation problem.

In [ ]:
class Net(nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        self.fc1 = nn.Linear(28 * 28, 128)
        self.fc2 = nn.Linear(128, 64)
        self.fc3 = nn.Linear(64, 10)

    def forward(self, x):
        x = x.view(-1, 28 * 28)
        x = torch.relu(self.fc1(x))
        x = torch.relu(self.fc2(x))
        x = self.fc3(x)
        return torch.log_softmax(x, dim=1)

model = Net()

In [ ]:
criterion = nn.CrossEntropyLoss()
optimizer = optim.SGD(model.parameters(), lr=0.01)

Training the model using SGD optimizer with Learning rate = 0.01

In [ ]:
num_epochs = 25

for epoch in range(num_epochs):
    for batch_idx, (data, target) in enumerate(train_loader):
        optimizer.zero_grad()
        output = model(data)
        loss = criterion(output, target)
        loss.backward()
        optimizer.step()

        if batch_idx % 1000 == 0:
            print(f'Epoch {epoch}, Loss: {loss.item()}')

Epoch 0, Loss: 2.311805486679077
Epoch 1, Loss: 0.6606583595275879
Epoch 2, Loss: 0.3945094347000122
Epoch 3, Loss: 0.43225768208503723
Epoch 4, Loss: 0.2439851015806198
Epoch 5, Loss: 0.22991664707660675
Epoch 6, Loss: 0.23388774693012238
Epoch 7, Loss: 0.24321424961090088
Epoch 8, Loss: 0.22755777835845947
Epoch 9, Loss: 0.3205982446670532
Epoch 10, Loss: 0.1678895354270935
Epoch 11, Loss: 0.16380038857460022
Epoch 12, Loss: 0.2488624006509781
Epoch 13, Loss: 0.2240212857723236
Epoch 14, Loss: 0.12558647990226746
Epoch 15, Loss: 0.10311552882194519
Epoch 16, Loss: 0.1514860987663269
Epoch 17, Loss: 0.13634489476680756
Epoch 18, Loss: 0.0776069387793541
Epoch 19, Loss: 0.11640062183141708
Epoch 20, Loss: 0.07778479158878326
Epoch 21, Loss: 0.16256535053253174
Epoch 22, Loss: 0.07168073952198029
Epoch 23, Loss: 0.13237303495407104
Epoch 24, Loss: 0.09695573151111603


Evaluating accuracy of this ANN over the MNIST test dataset

In [ ]:
model.eval()
test_loss = 0
correct = 0

with torch.no_grad():
    for data, target in test_loader:
        output = model(data)
        test_loss += criterion(output, target).item()
        pred = output.argmax(dim=1)
        correct += pred.eq(target).sum().item()

test_loss /= len(test_loader.dataset)
accuracy = 100.0 * correct / len(test_loader.dataset)

print(f'Test Loss: {test_loss}, Accuracy: {accuracy}%')

Test Loss: 0.001230151367932558, Accuracy: 96.39%


Part 2 : MNIST Classification using AIHWKIT

Same Network definition, but this time using Hardware-aware Analog Linear layers

In [ ]:
import aihwkit.nn as aihwkit_nn
from aihwkit.optim import AnalogSGD
from aihwkit.simulator.configs import SingleRPUConfig
from aihwkit.simulator.configs.devices import ConstantStepDevice

class AnalogNet(nn.Module):
    def __init__(self, device):
        super().__init__()
        self.device = device
        self.config =  SingleRPUConfig(device = self.device)
        self.fc1 = aihwkit_nn.AnalogLinear(28 * 28, 128, rpu_config = self.config)
        self.fc2 = aihwkit_nn.AnalogLinear(128, 64, rpu_config = self.config)
        self.fc3 = aihwkit_nn.AnalogLinear(64, 10, rpu_config = self.config)

    def forward(self, x):
        x = x.view(-1, 28 * 28)  # Flatten the input
        x = torch.relu(self.fc1(x))
        x = torch.relu(self.fc2(x))
        x = self.fc3(x)
        return torch.log_softmax(x, dim=1)

Setting up all the presets

In [ ]:
from aihwkit.simulator.presets import ReRamSBPresetDevice, ReRamESPresetDevice,  CapacitorPresetDevice, IdealizedPresetDevice, PCMPresetDevice
presets = [ConstantStepDevice(w_min=-0.4), ReRamSBPresetDevice(), ReRamESPresetDevice(), CapacitorPresetDevice(), IdealizedPresetDevice(), PCMPresetDevice()]
device_names = ["ConstantStepDevice", "ReRamSBPresetDevice", "ReRamESPresetDevice", "CapacitorPresetDevice", "IdealizedPresetDevice", "PCMPresetDevice"]

In [ ]:
index = 0
num_epochs = 10
for device in presets:
  # Setting up the preset device configs and defining the optimizer
  analog_model = AnalogNet(device)
  optimizer = AnalogSGD(analog_model.parameters(), lr=0.01, momentum=0.5)
  optimizer.regroup_param_groups(analog_model)
  print("Experimenting with", device_names[index])
  index += 1

  # Training loop
  analog_model.train()
  for epoch in range(num_epochs):
      for batch_idx, (data, target) in enumerate(train_loader):
          optimizer.zero_grad()
          output = analog_model(data)
          loss = criterion(output, target)
          loss.backward()
          optimizer.step()

          if batch_idx % 1000 == 0:
              print(f'Epoch {epoch}, Loss: {loss.item()}')

  # Test loop
  analog_model.eval()
  test_loss = 0
  correct = 0
  with torch.no_grad():
      for data, target in test_loader:
          output = analog_model(data)
          test_loss += criterion(output, target).item()
          pred = output.argmax(dim=1)
          correct += pred.eq(target).sum().item()

  test_loss /= len(test_loader.dataset)
  final_an_accuracy = 100.0 * correct / len(test_loader.dataset)
  print("Final Validation Accuracy (Hardware-aware):", final_an_accuracy)

Experimenting with ConstantStepDevice
Epoch 0, Loss: 2.3017730712890625
Epoch 1, Loss: 0.6082190871238708
Epoch 2, Loss: 0.2916443347930908
Epoch 3, Loss: 0.3476111590862274
Epoch 4, Loss: 0.3610057532787323
Epoch 5, Loss: 0.42415955662727356
Epoch 6, Loss: 0.36414918303489685
Epoch 7, Loss: 0.459450900554657
Epoch 8, Loss: 0.5076990723609924
Epoch 9, Loss: 0.4227752387523651
Final Validation Accuracy (Hardware-aware): 81.25
Experimenting with ReRamSBPresetDevice
Epoch 0, Loss: 4.291841983795166
Epoch 1, Loss: 2.313577175140381
Epoch 2, Loss: 2.2925162315368652
Epoch 3, Loss: 2.2960031032562256
Epoch 4, Loss: 2.3212697505950928
Epoch 5, Loss: 2.3219447135925293
Epoch 6, Loss: 2.300779342651367
Epoch 7, Loss: 2.2879981994628906
Epoch 8, Loss: 2.3105103969573975
Epoch 9, Loss: 2.3013083934783936
Final Validation Accuracy (Hardware-aware): 12.44
Experimenting with ReRamESPresetDevice
Epoch 0, Loss: 4.018839359283447
Epoch 1, Loss: 2.312469244003296
Epoch 2, Loss: 2.3097116947174072
Epoch 

Final Validation Accuracies

| Device/Model | Accuracy |
|-----------------|-----------------|
| Pytorch   | 96.39%   |
| ConstantStepDevice   | 81.25%   |
| ReRamSBPresetDevice  | 12.44%   |
| ReRamESPresetDevice   | 11.34%   |
| CapacitorPresetDevice   | 74.36%  |
| IdealizedPresetDevice   | 90.79%  |
| PCMPresetDevice   | 11.35%  |


As expected, our ANN has peforms much better when compared to the other Hardware-aware devices.

In this assignment, we have succesfully used aihwkit, which is a useful tool that allows us to simulate hardware-aware neural networks and benchmark their performance, to compare the performance of various types of hardware aware neural networks on the popular MNIST classification task.